activerecord/lib/active_record/connection_adapters/abstract_adapter.rb

```diff
@@ -143,19 +142,7 @@ def replica?
       # Returns true if the connection is a replica, or if +prevent_writes+
       # is set to true.
       def preventing_writes?
-        replica? || prevent_writes
-      end
-
-      # Prevent writing to the database regardless of role.
-      #
-      # In some cases you may want to prevent writes to the database
-      # even if you are on a database that can write. `while_preventing_writes`
-      # will prevent writes to the database for the duration of the block.
-      def while_preventing_writes
-        original, @prevent_writes = @prevent_writes, true
-        yield
-      ensure
-        @prevent_writes = original
+        replica? || ActiveRecord::Base.connection_handler.prevent_writes
```

Comment:
I never remember this but, does each model have its different connection handler or all models always share the `ActiveRecord::Base` connection handler?

Reply:
Each type of connection has it's own handler depending on what you put in `connects_to`.

Given the following:

```ruby
class ApplicationRecord < ActiveRecord::Base
  self.abstract_class = true

  connects_to database: { writing: :primary, reading: :primary_replica }
end

class Person < ApplicationRecord
end

class AnimalsBase < ApplicationRecord
  self.abstract_class = true

  connects_to database: { writing: :animals, reading: :animals_replica }
end

class Dog < AnimalsBase
end
```

This will create 2 handlers a `:writing` handler and a `:reading` handler (aka role). Each of these handlers has 2 connections: one for application record (person) and one for animals base (dog). Because we use `connected_to` to swap the handler out:

```
ActiveRecord::Base.connected_to(role: :reading) do
  # do something on reading handler
end
```

means that `ActiveRecord::Base.connection_handler.prevent_writes` will ask the handler whether it should prevent writes instead of the connection. Since `connected_to` operates on the handler we can be confident it's passing the value down correctly.

Reply:
I added a test that demonstrates this is working correctly here https://github.com/rails/rails/pull/36469/files#diff-3ad790de5c6c5fbd6d956e443bcc0490R1579